modules/c++/mem/include/mem/ScratchMemory.h

```diff
@@ -0,0 +1,140 @@
+#include <map>
+#include <sstream>
+#include <string>
+#include <utility>
+#include <vector>
+#include <except/Exception.h>
+#include <mem/BufferView.h>
+#include <sys/Conf.h>
+
+namespace mem
+{
+
+/*!
+ *  \class ScratchMemory
+ *  \brief Handle reservation of scratch memory segments within a single buffer.
+ *
+ *  Memory segments should be reserved during a "setup" phase using the put
+ *  method. Once all segments are reserved, call the setup method to set up
+ *  the underlying memory and ensure the alignment requirements of each segment.
+ *  The get method may be used afterwards to obtain pointers to the memory
+ *  segments.
+ */
+class ScratchMemory
+{
+public:
+    ScratchMemory() :
+        mNumBytesNeeded(0),
+        mIsSetup(false)
+    {
+    }
+
+    /*!
+     * \brief Reserve a buffer segment within this scratch memory buffer.
+     *
+     * \param key Identifier for scratch segment
+     * \param numBytes Size of scratch buffer
+     * \param numBuffers Number of distinct buffers to set up
+     * \param alignment Number of bytes to align segment pointer
+     *
+     * \throws except::Exception if the given key has already been used
+     */
+    template <typename T>
+    void put(const std::string& key,
+             size_t numBytes,
+             size_t numBuffers = 1,
+             size_t alignment = sys::SSE_INSTRUCTION_ALIGNMENT);
+
+    /*!
+     * \brief Get pointer to buffer segment.
+     *
+     * \param key Identifier for scratch segment
+     * \param indexBuffer Index of distinct buffer
+     *
+     * \return Pointer to buffer segment
+     *
+     * \throws except::Exception if the scratch memory has not been set up,
+     *         the key does not exist, or index of buffer is out of bounds
+     */
+    template <typename T>
+    T* get(const std::string& key, size_t indexBuffer = 0);
+
+    /*!
+     * \brief Ensure underlying memory is properly set up and position segment
+     *        pointers.
+     *
+     * \param scratchBuffer Storage to use for scratch memory. If scratchBuffer
+     *        of size 0 is passed, memory is allocated internally.
+     *
+     * \throws except::Exception if the scratchBuffer passed in is too small
+     *         to hold the requested scratch memory or has size > 0 with null
+     *         data pointer
+     */
+    void setup(const mem::BufferView<sys::ubyte>& scratchBuffer =
+            mem::BufferView<sys::ubyte>());
+
+    /*!
+     * \brief Get number of bytes needed to store scratch memory, including the
+     *        maximum possible alignment overhead.
+     */
+    size_t getNumBytes() const
+    {
+        return mNumBytesNeeded;
+    }
+
+private:
+    ScratchMemory(const ScratchMemory&);
+    ScratchMemory& operator=(const ScratchMemory&);
+
+    struct Segment
+    {
+        Segment(size_t numBytes, size_t numBuffers, size_t alignment) :
+            numBytes(numBytes),
+            numBuffers(numBuffers),
+            alignment(alignment)
+        {
```

Comment:
Move the implementation of the constructor to the cpp file.

Reply:
Done.